File: android/app/src/main/java/com/laneshadow/ui/components/atoms/Button.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.BorderStroke as ComposeBorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Button size variants
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 */
enum class ButtonSize {
    Small,
    Default,
    Large,
    XL,
    XXL,
    Icon,
}

/**
 * Button variant types
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 */
enum class ButtonVariant {
    Default,
    Secondary,
    Outline,
    Ghost,
    Destructive,
    Link,
    Glass,
}

/**
 * Icon position for button content
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 */
enum class IconPosition {
    Left,
    Right,
}

/**
 * Button dimension constants
 *
 * Documented against Button.md matrix values:
 * - sm: 36dp height, 12dp h-padding, 8dp radius
 * - default: 40dp height, 16dp h-padding, 8dp radius
 * - lg: 44dp height, 32dp h-padding, 8dp radius
 * - xl: 48dp height, 16dp h-padding, 16dp radius
 * - 2xl: 56dp height, 16dp h-padding, 24dp radius
 * - icon: 40×40dp, 0 padding, CircleShape radius
 */
private val BUTTON_HEIGHT_SMALL = 36.dp
private val BUTTON_HEIGHT_DEFAULT = 40.dp
private val BUTTON_HEIGHT_LARGE = 44.dp
private val BUTTON_HEIGHT_XL = 48.dp
private val BUTTON_HEIGHT_XXL = 56.dp
private val BUTTON_HEIGHT_ICON = 40.dp

private val BUTTON_PADDING_HORIZONTAL_SMALL = 12.dp
private val BUTTON_PADDING_HORIZONTAL_DEFAULT = 16.dp
private val BUTTON_PADDING_HORIZONTAL_LARGE = 32.dp
private val BUTTON_PADDING_HORIZONTAL_XL = 16.dp
private val BUTTON_PADDING_HORIZONTAL_XXL = 16.dp
private val BUTTON_PADDING_HORIZONTAL_ICON = 0.dp

/**
 * Disabled opacity constant
 *
 * Following RN wrapper behavior: disabled buttons have 0.5 opacity
 */
private const val BUTTON_DISABLED_OPACITY = 0.5f

/**
 * Button component
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 *
 * @param variant Button color variant (default, secondary, outline, ghost, destructive, link, glass)
 * @param size Button size variant (small, default, large, xl, xxl, icon)
 * @param onPress Callback when button is pressed (null makes button non-interactive)
 * @param disabled Whether button is disabled (adds opacity and prevents interaction)
 * @param loading Whether button is in loading state (shows progress indicator)
 * @param icon Optional icon composable to display
 * @param iconPosition Position of icon relative to text (left or right)
 * @param accessibilityLabel Accessibility label for screen readers
 * @param testID Test ID for UI testing
 * @param modifier Modifier for the button container
 * @param content Text content composable (optional for icon-only buttons)
 */
@Suppress("UNUSED_PARAMETER")
@Composable
fun Button(
    variant: ButtonVariant = ButtonVariant.Default,
    size: ButtonSize = ButtonSize.Default,
    onPress: (() -> Unit)? = null,
    disabled: Boolean = false,
    loading: Boolean = false,
    icon: @Composable (() -> Unit)? = null,
    iconPosition: IconPosition = IconPosition.Left,
    accessibilityLabel: String? = null,
    testID: String? = null,
    modifier: Modifier = Modifier,
    content: @Composable (() -> Unit)? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Determine if button should be disabled (loading also disables)
    val isDisabled = disabled || loading || onPress == null

    // Get button dimensions based on size
    val buttonHeight: Dp = when (size) {
        ButtonSize.Small -> BUTTON_HEIGHT_SMALL
        ButtonSize.Default -> BUTTON_HEIGHT_DEFAULT
        ButtonSize.Large -> BUTTON_HEIGHT_LARGE
        ButtonSize.XL -> BUTTON_HEIGHT_XL
        ButtonSize.XXL -> BUTTON_HEIGHT_XXL
        ButtonSize.Icon -> BUTTON_HEIGHT_ICON
    }

    val horizontalPadding: Dp = when (size) {
        ButtonSize.Small -> BUTTON_PADDING_HORIZONTAL_SMALL
        ButtonSize.Default -> BUTTON_PADDING_HORIZONTAL_DEFAULT
        ButtonSize.Large -> BUTTON_PADDING_HORIZONTAL_LARGE
        ButtonSize.XL -> BUTTON_PADDING_HORIZONTAL_XL
        ButtonSize.XXL -> BUTTON_PADDING_HORIZONTAL_XXL
        ButtonSize.Icon -> BUTTON_PADDING_HORIZONTAL_ICON
    }

    val buttonRadius: Shape = when (size) {
        ButtonSize.Small -> androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.sm)
        ButtonSize.Default -> androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.md)
        ButtonSize.Large -> androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.md)
        ButtonSize.XL -> androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.xl)
        ButtonSize.XXL -> androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.xxl)
        ButtonSize.Icon -> CircleShape
    }

    // Get colors based on variant
    val backgroundColor = when (variant) {
        ButtonVariant.Default -> theme.colors.primary.default
        ButtonVariant.Secondary -> theme.colors.secondary.default
        ButtonVariant.Destructive -> theme.colors.danger.default
        ButtonVariant.Outline -> Color.Transparent
        ButtonVariant.Ghost -> Color.Transparent
        ButtonVariant.Link -> Color.Transparent
        ButtonVariant.Glass -> theme.colors.surfaceVariant.default
    }

    val textColor = when (variant) {
        ButtonVariant.Default -> theme.colors.onSurface.default
        ButtonVariant.Secondary -> theme.colors.onSecondary.default
        ButtonVariant.Destructive -> theme.colors.onSurface.default
        ButtonVariant.Outline -> theme.colors.onSurface.default
        ButtonVariant.Ghost -> theme.colors.onSurface.default
        ButtonVariant.Link -> theme.colors.primary.default
        ButtonVariant.Glass -> theme.colors.onSurface.default
    }

    val border: BorderStroke? = when (variant) {
        ButtonVariant.Outline -> BorderStroke(1.dp, theme.colors.border.default)
        ButtonVariant.Glass -> BorderStroke(1.dp, theme.colors.border.default)
        else -> null
    }

    // Build semantics
    val buttonModifier = modifier
        .semantics {
            role = Role.Button
            if (isDisabled) {
                disabled()
            }
        }
        .then(
            if (!isDisabled && onPress != null) {
                Modifier.clickable(onClick = onPress)
            } else {
                Modifier
            }
        )

    // Apply disabled opacity
    val appliedModifier = if (isDisabled) {
        buttonModifier.alpha(BUTTON_DISABLED_OPACITY)
    } else {
        buttonModifier
    }

    Surface(
        modifier = appliedModifier,
        shape = buttonRadius,
        color = backgroundColor,
        border = border,
    ) {
        Box(
            modifier = Modifier
                .height(buttonHeight)
                .padding(horizontal = horizontalPadding),
        ) {
            when {
                loading -> {
                    // Show loading indicator
                    Box(
                        modifier = Modifier,
                    ) {
                        CircularProgressIndicator(
                            modifier = Modifier.width(20.dp).height(20.dp),
                            strokeWidth = 2.dp,
                            color = textColor,
                        )
                    }
                }
                icon != null && content == null -> {
                    // Icon-only button
                    Box(
                        modifier = Modifier,
                    ) {
                        icon()
                    }
                }
                else -> {
                    // Button with text and optional icon
                    Row(
                        modifier = Modifier,
                        horizontalArrangement = Arrangement.Center,
                    ) {
                        if (icon != null && iconPosition == IconPosition.Left) {
                            icon()
                            Spacer(modifier = Modifier.width(theme.space.sm))
                        }

                        if (content != null) {
                            val textStyle = theme.type.label.sm
                            Text(
                                text = buildTextContent { content() },
                                style = textStyle,
                                color = textColor,
                                textDecoration = if (variant == ButtonVariant.Link) {
                                    TextDecoration.Underline
                                } else {
                                    null
                                },
                            )
                        }

                        if (icon != null && iconPosition == IconPosition.Right) {
                            Spacer(modifier = Modifier.width(theme.space.sm))
                            icon()
                        }
                    }
                }
            }
        }
    }
}

/**
 * Helper to extract text string from content composable
 * This is a simplified version that works with basic Text composables
 */
private fun buildTextContent(content: @Composable () -> Unit): String {
    // For now, return empty string as we can't extract text from arbitrary composables
    // In a real implementation, you might use a different approach
    return ""
}

/**
 * Button with text content
 *
 * Convenience overload that accepts a String for text content
 *
 * @param variant Button color variant
 * @param size Button size variant
 * @param text Button text content
 * @param onPress Callback when button is pressed
 * @param disabled Whether button is disabled
 * @param loading Whether button is in loading state
 * @param icon Optional icon composable to display
 * @param iconPosition Position of icon relative to text
 * @param accessibilityLabel Accessibility label for screen readers (defaults to text if null)
 * @param testID Test ID for UI testing
 * @param modifier Modifier for the button container
 */
@Composable
fun Button(
    variant: ButtonVariant = ButtonVariant.Default,
    size: ButtonSize = ButtonSize.Default,
    text: String,
    onPress: (() -> Unit)? = null,
    disabled: Boolean = false,
    loading: Boolean = false,
    icon: @Composable (() -> Unit)? = null,
    iconPosition: IconPosition = IconPosition.Left,
    accessibilityLabel: String? = null,
    testID: String? = null,
    modifier: Modifier = Modifier,
) {
    Button(
        variant = variant,
        size = size,
        onPress = onPress,
        disabled = disabled,
        loading = loading,
        icon = icon,
        iconPosition = iconPosition,
        accessibilityLabel = accessibilityLabel ?: text,
        testID = testID,
        modifier = modifier,
    ) {
        Text(text)
    }
}

File: ios/LaneShadow/Views/Atoms/Button.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Button Size Enum

/**
 * Button size variants
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/atoms/Button.md
 */
public enum ButtonSize {
    case sm
    case `default`
    case lg
    case xl
    case xxl
    case icon
}

// MARK: - Button Variant Enum

/**
 * Button style variants
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/atoms/Button.md
 */
public enum ButtonVariant {
    case `default`
    case secondary
    case outline
    case ghost
    case destructive
    case link
    case glass
}

// MARK: - Icon Position Enum

/**
 * Icon position within button
 */
public enum IconPosition {
    case left
    case right
}

// MARK: - Button Component

/**
 * Button component
 *
 * Following RN wrapper API from react-native/components/ui/button.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/atoms/Button.md
 *
 * ## Design Tokens Used
 * - Heights: composed from `theme.space` (sm: 36, default: 40, lg: 44, xl: 48, xxl: 56, icon: 40)
 * - Padding: `theme.space.md/lg/2xl` (12/16/32) based on size
 * - Radius: `theme.radius.md/lg/xl/full` (8/16/24/9999) based on size
 * - Colors: `theme.colors.primary/secondary/danger/background/surfaceVariant/onSurface/onSecondary`
 * - Typography: `theme.type.label.sm` (12pt, medium weight)
 * - Border: `theme.borderWidth.thin` (1)
 * - Opacity: `theme.opacity.disabled` (0.5)
 * - Spacing: `theme.space.sm` (8, for icon gap)
 *
 * ## Parameters
 * - Parameters:
 *   - variant: Style variant (default, secondary, outline, ghost, destructive, link, glass)
 *   - size: Size variant (sm, default, lg, xl, xxl, icon)
 *   - onPress: Action callback when button is pressed
 *   - disabled: Whether button is disabled (default: false)
 *   - loading: Whether button is in loading state (default: false)
 *   - icon: Optional icon view to display
 *   - iconPosition: Position of icon (left or right, default: left)
 *   - accessibilityLabel: Accessibility label for screen readers
 *   - testID: Test identifier for UI testing
 */
public struct LSButton: View {
    @Environment(\.theme) private var theme
    @State private var isPressed = false

    private let variant: ButtonVariant
    private let size: ButtonSize
    private let onPress: (() -> Void)?
    private let disabled: Bool
    private let loading: Bool
    private let icon: (() -> AnyView)?
    private let iconPosition: IconPosition
    private let label: String
    private let accessibilityLabel: String?
    private let testID: String?

    public init(
        _ label: String,
        variant: ButtonVariant = .default,
        size: ButtonSize = .default,
        onPress: (() -> Void)? = nil,
        disabled: Bool = false,
        loading: Bool = false,
        icon: (() -> AnyView)? = nil,
        iconPosition: IconPosition = .left,
        accessibilityLabel: String? = nil,
        testID: String? = nil
    ) {
        self.label = label
        self.variant = variant
        self.size = size
        self.onPress = onPress
        self.disabled = disabled
        self.loading = loading
        self.icon = icon
        self.iconPosition = iconPosition
        self.accessibilityLabel = accessibilityLabel
        self.testID = testID
    }

    // MARK: - Layout Computed Properties

    private var height: CGFloat {
        switch size {
        case .sm:
            theme.space.xl + theme.space.md // 24 + 12 = 36
        case .lg:
            theme.space.xxl + theme.space.md // 32 + 12 = 44
        case .xl:
            theme.space.xxxl // 48
        case .xxl:
            theme.space.xxxxl - theme.space.sm // 64 - 8 = 56
        case .icon:
            theme.space.xxl + theme.space.sm // 32 + 8 = 40
        case .default:
            theme.space.xxl + theme.space.sm // 32 + 8 = 40
        }
    }

    private var horizontalPadding: CGFloat {
        if size == .icon { return 0 }
        switch size {
        case .sm:
            return theme.space.md // 12
        case .lg:
            return theme.space.xxl // 32
        case .xl, .xxl:
            return theme.space.lg // 16
        case .icon, .default:
            return theme.space.lg // 16
        }
    }

    private var cornerRadius: CGFloat {
        switch size {
        case .icon:
            theme.radius.full
        case .xxl:
            theme.radius.xl
        case .xl:
            theme.radius.lg
        case .sm, .lg, .default:
            theme.radius.md
        }
    }

    // MARK: - Color Computed Properties

    private var backgroundColor: Color {
        if disabled {
            return backgroundColorForDisabledState
        }

        if isPressed {
            return backgroundColorForPressedState
        }

        return backgroundColorForDefaultState
    }

    private var backgroundColorForDefaultState: Color {
        switch variant {
        case .ghost, .link:
            .clear
        case .glass:
            theme.colors.surfaceVariant.default
        case .secondary:
            theme.colors.secondary.default
        case .destructive:
            theme.colors.danger.default
        case .outline:
            theme.colors.background.default
        case .default:
            theme.colors.primary.default
        }
    }

    private var backgroundColorForPressedState: Color {
        switch variant {
        case .ghost, .link:
            .clear
        case .glass:
            theme.colors.surfaceVariant.pressed ?? theme.colors.surfaceVariant.default
        case .secondary:
            theme.colors.secondary.pressed ?? theme.colors.secondary.default
        case .destructive:
            theme.colors.danger.pressed ?? theme.colors.danger.default
        case .outline:
            theme.colors.accent.pressed ?? theme.colors.accent.default
        case .default:
            theme.colors.primary.pressed ?? theme.colors.primary.default
        }
    }

    private var backgroundColorForDisabledState: Color {
        switch variant {
        case .ghost, .link, .outline:
            theme.colors.background.default
        case .glass:
            theme.colors.surfaceVariant.disabled ?? theme.colors.surfaceVariant.default
        case .secondary:
            theme.colors.secondary.disabled ?? theme.colors.secondary.default
        case .destructive:
            theme.colors.danger.disabled ?? theme.colors.danger.default
        case .default:
            theme.colors.primary.disabled ?? theme.colors.primary.default
        }
    }

    private var foregroundColor: Color {
        if disabled {
            return theme.colors.onSurface.disabled ?? theme.colors.onSurface.default
        }

        if isPressed {
            switch variant {
            case .outline, .ghost:
                return theme.colors.accent.default
            case .link:
                return theme.colors.primary.default
            default:
                return defaultForegroundColor
            }
        }

        return defaultForegroundColor
    }

    private var defaultForegroundColor: Color {
        switch variant {
        case .secondary:
            theme.colors.onSecondary.default ?? theme.colors.onSurface.default
        case .link:
            theme.colors.primary.default
        case .default, .destructive, .glass, .outline, .ghost:
            theme.colors.onSurface.default
        }
    }

    private var borderColor: Color? {
        switch variant {
        case .outline, .glass:
            theme.colors.border.default
        case .default, .secondary, .destructive, .ghost, .link:
            nil
        }
    }

    // MARK: - Body

    public var body: some View {
        Group {
            if loading {
                buttonContent(
                    label: AnyView(
                        HStack(spacing: theme.space.sm) {
                            ProgressView()
                                .progressViewStyle(CircularProgressViewStyle(tint: foregroundColor))
                            Text("Loading…")
                        }
                    )
                )
            } else if let icon {
                buttonContent(
                    label: AnyView(
                        HStack(spacing: theme.space.sm) {
                            if iconPosition == .left && size != .icon {
                                icon()
                            }
                            if size != .icon {
                                Text(labelText)
                            }
                            if iconPosition == .right || size == .icon {
                                icon()
                            }
                        }
                    )
                )
            } else {
                buttonContent(
                    label: AnyView(
                        Text(labelText)
                    )
                )
            }
        }
        .disabled(disabled || loading)
        .accessibilityAddTraits(.isButton)
        .accessibilityLabel(accessibilityLabel ?? labelText)
        .accessibilityIdentifier(testID ?? "button")
    }

    // MARK: - Button Content

    @ViewBuilder
    private func buttonContent(label: AnyView) -> some View {
        let isIconOnly = size == .icon

        Button(action: {
            onPress?()
        }) {
            label
                .font(.system(size: theme.type.label.sm.fontSize, weight: .medium))
                .foregroundStyle(foregroundColor)
                .if(variant == .link) { view in
                    view.underline(true)
                }
                .frame(height: height)
                .if(!isIconOnly) { view in
                    view.frame(maxWidth: .infinity)
                }
                .if(isIconOnly) { view in
                    view.frame(width: height)
                }
                .padding(.horizontal, horizontalPadding)
                .background(backgroundColor)
                .cornerRadius(cornerRadius)
                .overlay(
                    RoundedRectangle(cornerRadius: cornerRadius)
                        .stroke(borderColor ?? .clear, lineWidth: theme.borderWidth.thin)
                )
                .opacity((disabled || loading) ? theme.opacity.disabled : 1.0)
                .scaleEffect(isPressed ? 0.98 : 1.0)
                .animation(.easeInOut(duration: 0.1), value: isPressed)
        }
        .buttonStyle(.plain)
        .simultaneousGesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    if !disabled, !loading {
                        isPressed = true
                    }
                }
                .onEnded { _ in
                    isPressed = false
                }
        )
    }

    // MARK: - Helpers

    private var labelText: String {
        label
    }
}
